expr/functions/hitcount/function.go

```diff
@@ -95,41 +99,58 @@ func (f *hitcount) Do(ctx context.Context, e parser.Expr, from, until int64, val
 		}
 		r.Tags["hitcount"] = fmt.Sprintf("%d", bucketSizeInt32)
 
-		bucketEnd := start + bucketSize
-		t := arg.StartTime
-		ridx := 0
-		var count float64
-		bucketItems := 0
-		for _, v := range arg.Values {
-			bucketItems++
-			if !math.IsNaN(v) {
-				if math.IsNaN(count) {
-					count = 0
-				}
+		step := arg.StepTime
+		buckets := make([][]float64, bucketCount)
+		newStart := stop - bucketCount*interval
+		r.StartTime = newStart
 
-				count += v * float64(arg.StepTime)
+		for i, v := range arg.Values {
+			if math.IsNaN(v) {
+				continue
 			}
 
-			t += arg.StepTime
+			start_time := arg.StartTime + int64(i)*step
+			startBucket, startMod := helper.Divmod(start_time-newStart, interval)
+			end_time := start_time + step
+			endBucket, endMod := helper.Divmod(end_time-newStart, interval)
 
-			if t >= stop {
-				break
+			if endBucket >= bucketCount {
+				endBucket = bucketCount - 1
+				endMod = interval
 			}
 
-			if t >= bucketEnd {
-				r.Values[ridx] = count
-
-				ridx++
-				bucketEnd += bucketSize
-				count = math.NaN()
-				bucketItems = 0
+			if startBucket == endBucket {
+				// All hits go into a single bucket
+				if startBucket >= 0 {
+					buckets[startBucket] = append(buckets[startBucket], v*float64(endMod-startMod))
+				}
+			} else {
+				// Spread the hits amongst 2 or more buckets
+				if startBucket >= 0 {
+					buckets[startBucket] = append(buckets[startBucket], v*float64(interval-startMod))
+				}
+				hitsPerBucket := v * float64(interval)
+				for j := startBucket + 1; j < endBucket; j++ {
+					buckets[j] = append(buckets[j], hitsPerBucket)
+				}
+				if endMod > 0 {
+					buckets[endBucket] = append(buckets[endBucket], v*float64(endMod))
+				}
 			}
 		}
-
-		// remaining values
-		if bucketItems > 0 {
-			r.Values[ridx] = count
+		newValues := make([]float64, bucketCount)
+		for i, bucket := range buckets {
+			if bucket != nil {
```

Comment:
`bucket` should never be `nil` AFAICT.  I think what [graphite web is doing](https://github.com/graphite-project/graphite-web/blob/master/webapp/graphite/render/functions.py#L5509) is checking if the bucket is empty, so:

```suggestion
			if len(bucket) == 0 {
```